test/langtools/tools/javac/patterns/TypedDeconstructionPatternExc.java

```diff
@@ -83,6 +86,51 @@ int testExprCond(Pair<String, Integer> p) {
         }
     }
 
+    void testTryExpr() {
+        TEST: {
+            try {
+                var v = switch ((Pair<String, Integer>) (Object) new Pair<Integer, Integer>(1, 1)) {
+                    case Pair<String, Integer>(String s, Integer i) -> s.length() + i;
+                    case Object o -> -1;
+                };
+            } catch (ClassCastException ex) {
+                //OK
+                break TEST;
+            } catch (Throwable t) {
+                t.printStackTrace();
+                fail("Unexpected Throwable!");
+            }
+            fail("ClassCastException not thrown!");
+        }
+        TEST: {
+            try {
+                var v = switch (new Pair<String, Integer>("fail", 1)) {
+                    case Pair<String, Integer>(String s, Integer i) -> s.length() + i;
+                    case Object o -> -1;
+                };
+            } catch (MatchException ex) {
+                //OK
+                break TEST;
+            } catch (Throwable t) {
+                t.printStackTrace();
+                fail("Unexpected Throwable!");
+            }
+            fail("MatchException not thrown!");
+        }
+    }
+
+    int testLambda(Pair<String, Integer> p) {
+        var r = prepareLambda();
+        return r.applyAsInt(p);
+    }
+
+    ToIntFunction<Pair<String, Integer>> prepareLambda() {
+        return p -> switch (p) {
+            case Pair<String, Integer>(String s, Integer i) -> s.length() + i;
+            case Object o -> -1;
+        };
+    }
+
     static final String EXCEPTION_MESSAGE = "exception-message";
 
     record Pair<L, R>(L l, R r) {
```

Comment:
I think that for completeness we need records with generic record components, unless those are not accepted yet as record patterns